### PR TITLE
Introduce TORCH_DISABLE_GPU_ASSERTS

### PR DESCRIPTION
**Pull request by pruthvistony**
- Asserts for CUDA are enabled by default
- Disabled for ROCm by default by setting `TORCH_DISABLE_GPU_ASSERTS` to `ON`
- Can be enabled for ROCm by setting above variable to`OFF` during build or can be forcefully enabled by setting `ROCM_FORCE_ENABLE_GPU_ASSERTS:BOOL=ON`

This is follow up changes as per comment in PR #81790, comment [link](https://github.com/pytorch/pytorch/pull/81790#issuecomment-1215929021)

### CODE REVIEW

**Comment by pruthvistony** [comment]
The 1 failing job was due to timing out, so rebase-ing the patch to rerun all jobs.

**Comment by pruthvistony** [comment]
The job - [linux-bionic-py3.7-clang9 / test (dynamo, 1, 2, linux.2xlarge)](https://github.com/pytorch/pytorch/runs/8087176704?check_suite_focus=true#logs)
is failing due to unrelated error.

**Review by jeffdaily — APPROVED** [review]

**Review by jeffdaily — APPROVED** [review]

**Comment by pruthvistony** [comment]
@malfet ,
Can you please help review this PR.

**Comment by kit1980** [comment]
@pruthvistony Should we merge this?

**Comment by pruthvistony** [comment]
> @pruthvistony Should we merge this?

@kit1980,
As I discuss with @malfet, got to know that this PR can break internal builds.
But if the internal builds are fine, we can merge it. Please confirm on it.

**Comment by kit1980** [comment]
I'm seeing "error: invalid preprocessing directive #cmakedefine TORCH_ENABLE_GPU_ASSERTS" when running internally on Meta code. I need to investigate further to understand what to do.

**Comment by pruthvistony** [comment]
> I'm seeing "error: invalid preprocessing directive #cmakedefine TORCH_ENABLE_GPU_ASSERTS" when running internally on Meta code. I need to investigate further to understand what to do.

@kit1980 ,
Thanks for checking, I will also check on my side.

**Comment by pruthvistony** [comment]
@kit1980 @malfet ,
I have negated the kernel assert handling logic, please let me know if this PR builds fine on internal builds.

**Comment by pruthvistony** [comment]
> @kit1980 @malfet , I have negated the kernel assert handling logic, please let me know if this PR builds fine on internal builds.

@malfet @kit1980 ,
Please let me know if the internal builds are fine. Thank you.

**Review by malfet — APPROVED** [review]
Looks ok (though I'll need to tweak internal build system still)
Though from build-system/cmake point of view it looks very weird:
 - PR introduces `TORCH_DISABLE_GPU_ASSERTS` which is respected when CUDA is used, but ignores when ROCM
 - It also introduces `FORCE_ENABLE_GPU_ASSERTS`, which is respected when ROCM is used, but not when CUDA , but it does not have a ROCM prefix nor is checked under `if USE_ROCM` condition. 
 Not sure what is wrong with the existing `ROCM_ENABLE_KERNEL_ASSERTS` variable?

File: cmake/Dependencies.cmake [hunk 1]
```diff
@@ -1248,6 +1248,21 @@ if(ANDROID)
...
+  message(STATUS "Disabling kernel asserts for ROCm")
+  caffe2_update_option(TORCH_DISABLE_GPU_ASSERTS ON)
+else()
+  message(STATUS "Enabling kernel asserts")
+endif()
+
+if(FORCE_ENABLE_GPU_ASSERTS)
+  caffe2_update_option(TORCH_DISABLE_GPU_ASSERTS OFF)
+  message(STATUS "Forcefully enabling kernel asserts")
+endif()
```
Comment (malfet) [inline cmake/Dependencies.cmake:1264]: Wouldn't something like that be cleaner?
```suggestion
if(USE_ROCM AND NOT ROCM_FORCE_ENABLE_GPU_ASSERTS)
  message(STATUS "Disabling kernel asserts for ROCm")
  caffe2_update_option(TORCH_DISABLE_GPU_ASSERTS ON)
elseif(USE_ROCM AND ROCM_FORCE_ENABLE_GPU_ASSERTS)
  message(STATUS "Forcefully enabling kernel asserts on ROCM")
endif()
```

**Review by malfet — CHANGES_REQUESTED** [review]
Seem suggested changes (and please add `ROCM_` prefix to `FORCE_ENABLE` as it is a no-op for CUDA

File: cmake/Dependencies.cmake (hunk 1 above)
Comment (pruthvistony) [reply cmake/Dependencies.cmake:1264]: Tried to keep the variable  `FORCE_ENABLE_GPU_ASSERTS` generic, will keep it ROCm specific and update the code accordingly.

Comment (pruthvistony) [reply cmake/Dependencies.cmake]: Updated the code.

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
It should have been landed via co-dev as it still need internal changes. Reverting and reopening

**Comment by malfet** [comment]
@mehtanirav Internal copy of this PR landed a while back, not sure why ShipIt did not close OSS one

**Pull request closed without merging**